src/protocolspectator.cpp

```diff
@@ -0,0 +1,383 @@
+}
+
+void ProtocolSpectator::syncOpenContainers()
+{
+	const auto& openContainers = player->getOpenContainers();
+	for (const auto& it : openContainers) {
+		auto openContainer = it.second;
+		auto container = openContainer.container;
+		sendContainer(it.first, container, container->hasParent(), openContainer.index);
+	}
+}
+
+void ProtocolSpectator::login(const std::string& liveCastName, const std::string& password)
+{
+	//dispatcher thread
+	auto _player = g_game.getPlayerByName(liveCastName);
+	if (!_player || _player->isRemoved()) {
+		disconnectSpectator("Live cast no longer exists. Please relogin to refresh the list.");
+		return;
+	}
+
+	const auto liveCasterProtocol = ProtocolGame::getLiveCast(_player);
+
+	if (!liveCasterProtocol) {
+		disconnectSpectator("Live cast no longer exists. Please relogin to refresh the list.");
+		return;
+	}
+
+	const auto& liveCastPassword = liveCasterProtocol->getLiveCastPassword();
+	if (liveCasterProtocol->isLiveCaster()) {
+		if (!liveCastPassword.empty() && password != liveCastPassword) {
+			disconnectSpectator("Wrong live cast password.");
+			return;
+		}
+
+		player = _player;
+		player->incrementReferenceCounter();
+		eventConnect = 0;
+		client = liveCasterProtocol;
+		m_acceptPackets = true;
+		OutputMessagePool::getInstance().addProtocolToAutosend(shared_from_this());
+		sendAddCreature(player, player->getPosition(), 0, false);
+		syncKnownCreatureSets();
+		syncChatChannels();
+		syncOpenContainers();
+
+		liveCasterProtocol->addSpectator(std::static_pointer_cast<ProtocolSpectator>(shared_from_this()));
+	} else {
+		disconnectSpectator("Live cast no longer exists. Please relogin to refresh the list.");
+	}
+}
+
+void ProtocolSpectator::logout()
+{
+	m_acceptPackets = false;
+	if (client && player) {
+		client->removeSpectator(std::static_pointer_cast<ProtocolSpectator>(shared_from_this()));
+		player->decrementReferenceCounter();
+		player = nullptr;
+	}
+	disconnect();
+}
+
+void ProtocolSpectator::parsePacket(NetworkMessage& msg)
+{
+	if (!m_acceptPackets || g_game.getGameState() == GAME_STATE_SHUTDOWN || msg.getLength() <= 0) {
+		return;
+	}
+
+	uint8_t recvbyte = msg.getByte();
+
+	if (!player) {
+		if (recvbyte == 0x0F) {
+			disconnect();
+		}
+
+		return;
+	}
+
+	//a dead player can not perform actions
+	if (player->isRemoved() || player->getHealth() <= 0) {
+		disconnect();
+		return;
+	}
+
+	switch (recvbyte) {
+		case 0x14: g_dispatcher.addTask(createTask(std::bind(&ProtocolSpectator::logout, getThis()))); break;
+		case 0x1D: g_dispatcher.addTask(createTask(std::bind(&ProtocolSpectator::sendPingBack, getThis()))); break;
+		case 0x1E: g_dispatcher.addTask(createTask(std::bind(&ProtocolSpectator::sendPing, getThis()))); break;
+		//Reset viewed position/direction if the spectator tries to move in any way
+		case 0x65: case 0x66: case 0x67: case 0x68: case 0x69: case 0x6A: case 0x6B: case 0x6C: case 0x6D: case 0x6E: case 0x6F: case 0x70: case 0x71:
```

Comment:
Add 0x64 (AutoWalk) and remove 0x6E (isn't used). Also, 0x69 (StopAutoWalk) shouldn't be necessary here.